### DISCUSSION TITLE
Feature request: Translating JavaScript/TypeScript/Lua to V

### DISCUSSION THREAD

**Discussion by zaoqi** [vlang/v#11576] [Ideas]

**Comment by xland** [comment]
I think it is impossible~
*_*

**Comment by zaoqi** [comment]
@xland 爲何？

**Comment by GGXBoo** [comment]
这不需要，因为这语言定位就不是这些

**Comment by medvednikov** [comment]
Yes, it would be very nice to support as many languages as possible.

Right now C => V works very well. C++ => V is at an early stage, but I think that supporting the entire standard is doable.

Translating dynamic languages or languages with powerful runtimes (like Go) is a lot harder, but it's definitely not impossible.

Python => V would be extremely cool.

**Comment by cjmxp** [comment]
I don't think that's a good idea. For example, haxe compiles all kinds of languages, but its performance is poor. 
V should follow the golang route and make up for the deficiency of golang.

**Comment by medvednikov** [comment]
Haven't heard about Haxe, but looks like it translates to different languages.

Here we are talking about the opposite: translating other languages to V.

This will not make V slower.

**Comment by ntrel** [comment]
> C++ => V is at an early stage, but I think it's supporting the entire standard is doable.

C++ is very complex, with implicit user type conversions, function overloading, template function overloading, destructors, exceptions, two kinds of references etc. I'm finding it hard to imagine C++ code using these features being automatically ported to V in a readable format. Even for C, how would V handle global variables?

**Comment by medvednikov** [comment]
Globals are allowed in translated code. Along with some other features.

The resulting code is not very nice, but definitely readable. This will improve as the translator matures.

**Comment by superwhiskers** [comment]
> Globals are allowed in translated code. Along with some other features.

allowing globals in translated code only is a pretty big inconsistency (at least to me.)

this begs a few questions:
- how does `v` know if a specific source file is translated?
- couldn't this be abused to allow globals in regular code?

i personally think this feature should be removed in favor of either:
- allowing globals in all source files
- removing globals entirely
- or adding a kind of "strict mode" where globals (and potentially other features) are disabled

**Comment by nedpals** [comment]
Updates:

- JS backend is in the works. (See: https://github.com/vlang/v/tree/master/vlib/v/gen/js)
- A Typescript backend is pointless since the generated `.ts` file must be compiled into a javascript file in order to run.
- A Lua backend might be possible.
- ...and globals are only allowed when setting the compiler with  the `--enable-globals` flag

**Comment by ambyjkl** [comment]
> A Typescript backend is pointless

@nedpals I agreed. How about generating `.d.ts` files then?

**Comment by avecchio** [comment]
> Yes, it would be very nice to support as many languages as possible.
> 
> Right now C => V works very well. C++ => V is at an early stage, but I think that supporting the entire standard is doable.
> 
> Translating dynamic languages or languages with powerful runtimes (like Go) is a lot harder, but it's definitely not impossible.
> 
> Python => V would be extremely cool.

Facebook appears to have created a machine learning model to translate from one language into another. Does anyone have concerns on using such a model to translate 80%+ of the code correctly? https://www.infoq.com/news/2020/06/facebook-ai-transpiler/

**Comment by JalonSolov** [comment]
I don't if there's enough V code in existence to train it to a new model. Right now, it only supports a very few languages.  From their readme on Github:  `Transcompilation between Java, C++ and Python 3.`
